### PR TITLE
Add support for local trees mode in forest

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

Adds API to sklearnex to utilize changes added in https://github.com/uxlfoundation/oneDAL/pull/3139. Supports `local_trees_mode` parameter in Random Forest SPMD classes without breaking scikit-learn inheritor requirements. Adds test for this parameter (just functionality). Docs need to be updated somehow.

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/ensemble/forest.cpp [hunk 1]
```diff
@@ -201,6 +201,15 @@ struct params2desc {
         desc.set_seed(params["seed"].cast<std::int64_t>());
 #endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
 
+#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20250700
+        if (params.contains("local_trees_mode")) {
+            desc.set_local_trees_mode(params["local_trees_mode"].cast<bool>())
```
Comment (ethanglaser) [inline onedal/ensemble/forest.cpp]:
```suggestion
            desc.set_local_trees_mode(params["local_trees_mode"].cast<bool>());
```

File: onedal/ensemble/forest.cpp [hunk 2]
```diff
@@ -201,6 +201,15 @@ struct params2desc {
         desc.set_seed(params["seed"].cast<std::int64_t>());
 #endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
 
+#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20250700
+        if (params.contains("local_trees_mode")) {
+            desc.set_local_trees_mode(params["local_trees_mode"].cast<bool>());
+        }
+        else {
```
Comment (icfaust) [inline onedal/ensemble/forest.cpp]: https://github.com/uxlfoundation/oneDAL/blob/29a231fc695381f1e2c1d7f8fb21d029b4d65c5d/cpp/oneapi/dal/algo/decision_forest/common.cpp#L70  is else branch necessary?

File: sklearnex/spmd/ensemble/forest.py [hunk 3]
```diff
@@ -25,6 +26,115 @@ class RandomForestClassifier(RandomForestClassifier_Batch):
...
+                warm_start=warm_start,
+                class_weight=class_weight,
+                ccp_alpha=ccp_alpha,
+                max_samples=max_samples,
+                max_bins=max_bins,
+                min_bin_size=min_bin_size,
+            )
+
+    def _create_onedal_estimator(self, onedal_params):
+        onedal_params = dict(onedal_params)  # copy to avoid mutating input
```
Comment (icfaust) [inline sklearnex/spmd/ensemble/forest.py]: Whats the mutation avoidance for?

**Review by icfaust — COMMENTED** [review]
I'd go with a wrapper over ```_onedal_factory```. It would isolate the changes only to spmd. Something like
```python

def local_trees_wrapper(func):
    def new_factory(self, **params):
        params["local_trees_mode"] = self.local_trees_mode
        return func(self, **params)
    return new_factory
```

Additional things:
 - The addition of testing is 🔥 
 - Add some code comments in the spmd file so that the next person can understand why the init was added (since its so many parameters, it may be hard to spot).
 - Overall I am surprised how cleanly this came through.  Its definitely something for others to reference to going forward.

File: onedal/ensemble/forest.cpp [hunk 4]
```diff
@@ -201,6 +201,15 @@ struct params2desc {
...
+            desc.set_local_trees_mode(false);
+        }
```
Comment (ethanglaser) [inline onedal/ensemble/forest.cpp]:
```suggestion
```

File: sklearnex/spmd/ensemble/forest.py (hunk 3 above)
Comment (ethanglaser) [reply sklearnex/spmd/ensemble/forest.py]: Probably not necessary

Comment (ethanglaser) [inline sklearnex/spmd/ensemble/forest.py]:
```suggestion
```

File: sklearnex/spmd/ensemble/forest.py [hunk 5]
```diff
@@ -50,6 +159,111 @@ class RandomForestRegressor(RandomForestRegressor_Batch):
...
+                verbose=verbose,
+                warm_start=warm_start,
+                ccp_alpha=ccp_alpha,
+                max_samples=max_samples,
+                max_bins=max_bins,
+                min_bin_size=min_bin_size,
+            )
+
+    def _create_onedal_estimator(self, onedal_params):
+        onedal_params = dict(onedal_params)  # copy to avoid mutating input
```
Comment (ethanglaser) [inline sklearnex/spmd/ensemble/forest.py]:
```suggestion
```

**Comment by ethanglaser** [comment]
> I'd go with a wrapper over `_onedal_factory`. It would isolate the changes only to spmd. 

Great suggestion, didn't fully realize this was possible. Added.

**Comment by ethanglaser** [comment]
Green CI from before reformat: http://intel-ci.intel.com/f06d1269-bcb8-f1a2-8fdc-d4f5ef20c6a0

**Comment by ethanglaser** [comment]
@icfaust got local_trees_wrapper implemented such that no changes in non-spmd file are necessary (just returns class instead of function) and CI appears to be okay. Feel free to suggest tweaks to the updated implementation.

**Review by icfaust — APPROVED** [review]
that green CI is chefs kiss. thanks for the hard work it looks great
